### PR TITLE
feat: WhileLoopPrimitive for bounded graph execution in lax.scan

### LINKED ISSUE #134: `metal::malloc` resource limit crash on long `lax.scan` / `lax.while_loop`
### Code to reproduce

```python
import jax
import jax.numpy as jnp
from jax import random
def make_lstm_params(key, input_dim, hidden_dim):
    keys = random.split(key, 8)
    # Wi, Wf, Wc, Wo (input projections) + Ui, Uf, Uc, Uo (recurrent projections)
    return tuple(random.normal(k, (input_dim if i < 4 else hidden_dim, hidden_dim)) * 0.01
                 for i, k in enumerate(keys))
def lstm_step(carry, x):
    (h, c), params = carry
    Wi, Wf, Wc, Wo, Ui, Uf, Uc, Uo = params
    i = jax.nn.sigmoid(x @ Wi + h @ Ui)
    f = jax.nn.sigmoid(x @ Wf + h @ Uf)
    g = jnp.tanh(x @ Wc + h @ Uc)
    o = jax.nn.sigmoid(x @ Wo + h @ Uo)
    c_new = f * c + i * g
    h_new = o * jnp.tanh(c_new)
    return ((h_new, c_new), params), h_new
hidden = 64
params = make_lstm_params(random.key(0), 32, hidden)
xs = jnp.ones((10_000, 32))
def loss_fn(params):
    init = ((jnp.zeros(hidden), jnp.zeros(hidden)), params)
    _, ys = jax.lax.scan(lstm_step, init, xs)
    return ys.sum()
jax.grad(loss_fn)(params)  # crashes: metal::malloc resource limit exceeded
```

## Problem
Long-running `lax.scan` / `lax.while_loop` (e.g., LSTM fwd+bwd with ≥10k timesteps) crash with:
```
[MPS ERROR] Exception dispatching stablehlo.dynamic_update_slice: [metal::malloc] Resource limit (499000) exceeded.
[MPS ERROR] stablehlo.while: failed to execute body region
```
## Current state
`HandleWhile` throws `CompileIncompatibleError` when called inside `mx::compile`, forcing the entire outer compilation to fall back to the interpreted MLIR walker. In this path, each loop iteration accumulates MLX graph nodes without evaluation. For long loops, unbounded graph growth exceeds Metal's ~500k allocation limit.

## Approach (I am already preparing a PR)
Replace the `CompileIncompatibleError` throw with a custom `WhileLoopPrimitive` — an `mlx::core::Primitive` subclass that is opaque to `mx::compile` but executes the loop internally:
- loop body wrapped via `mx::compile()` for GPU execution
- The primitive runs on the CPU stream for orchestration only (this does not cause any issues as far as I can tell) — the compiled body dispatches GPU kernels via the default GPU stream. Running on the GPU stream would deadlock (`eval()` inside a GPU eval callback re-enters the GPU scheduler).

### Version and runtime information

```text
jax:     0.9.2
jaxlib:  0.9.2
jax-mps: 0.9.14
Python:  3.13.12
macOS:   26.3.1 (25D2128)
Chip:    Apple M3 Max
Memory:  48 GB
Backend: mps
```

### PR DESCRIPTION
**Pull request by tsumme1**
Add `WhileLoopPrimitive`, a custom MLX primitive that executes `stablehlo.while` loops with per-iteration `eval()` to bound graph depth. This prevents the `metal::malloc` resource limit crash that occurs when reverse-mode AD through `lax.scan` builds an unbounded computation graph.

Fixes #134, addresses #83.

### `WhileLoopPrimitive` (`control_flow.cc`)
- Custom `mlx::core::Primitive` that runs on the CPU stream, orchestrating compiled body+cond functions with per-step `eval()`.
- External values are threaded as explicit inputs (via `CollectExternalValues`) to avoid closure-capture issues with `mx::compile`.
- Body+cond compiled via the **public** `mlx::core::compile()` overload — the returned `std::function` owns the cache entry through a `shared_ptr` deleter that calls `compile_erase` automatically when the primitive's last copy dies. No manual cache bookkeeping or shutdown guard needed (matches the `MlxExecutable` pattern post-#146/#148).
- Defensive validation: CPU stream assertion before compilation, `condResults` arity check before indexing.

### `pjrt_mutex.h`
- Process-global recursive mutex for re-entrant `Execute()` calls during `WhileLoopPrimitive::eval_impl()`.

### `bench_rnn_training.py`
- GRU training benchmark exercising grad-through-scan.

### `tests/configs/control_flow.py`
- New `lax.scan.long_grad` regression config: 2000-iter scan with reverse-mode AD through a differentiable scalar, mirroring #134's pattern. Exercises the bounded-graph path under both jit and eager.

## Design

The primitive is transparent to `mx::compile` tracing — it appears as an opaque node whose `eval_impl` internally dispatches compiled GPU kernels. Nested while-loops are supported in the PJRT compile path (`allow_while_primitive=true`); the eager path falls back to uncompiled `ExecuteRegion` for nested cases.

## Testing

- All control-flow tests pass locally (MPS + CPU), including the new `lax.scan.long_grad` regression.
- Full local suite: 2137 passed, 232 skipped, 38 xfailed.
- `scripts/run_jax_tests.py` runs to completion.

## Rebase / cleanup notes

Branch was rebased onto current `main` (post-#146 / #147 / #148). Resolved one conflict in `mlx_executable.cc` by adopting `main`'s `~MlxExecutable() = default` (per #148, the detach-on-destroy patch was dropped). A follow-up cleanup commit then drops the now-redundant `IsProcessShuttingDown` / `CacheGuard` / `mlx::core::detail::compile` plumbing in favor of the public compile API, matching the rationale in #148.

### CODE REVIEW

File: src/pjrt_plugin/ops/handler_utils.h [hunk 1]
```diff
@@ -23,13 +25,37 @@
...
+// is reliably set before any shutdown-time destructors run.
+inline std::atomic<bool>& ProcessShutdownFlag() {
+    static std::atomic<bool> flag{false};
+    return flag;
+}
+
+inline const int kProcessShutdownHookInstalled = [] {
+    std::atexit([] { ProcessShutdownFlag().store(true, std::memory_order_relaxed); });
+    return 0;
+}();
```
[the remark this answers is not in the document]
Comment (tsumme1) [reply src/pjrt_plugin/ops/handler_utils.h]: I have softened the comment's statement.

**Comment by tsumme1** [comment]
@tillahoffmann Does this PR successfully run the whole JAX test suite on your device?

**Comment by tillahoffmann** [comment]
I'm still seeing the occasional hang in `tests/lax_numpy_indexing_test.py`, but I'll run the test suite a few times on `main` to see whether this is actually a pre-existing issue.

Something Claude highlighted that might be relevant:

> At process exit on the previous successful run, the stack trace was:
> 
> ```
>  exit → __cxa_finalize_ranges → CompilerCache::~CompilerCache()
>   → ... → mlx::core::array::~array()
>   → WhileLoopPrimitive::~WhileLoopPrimitive()
>   → shared_ptr<...>::__on_zero_shared()
>   → mlx::core::detail::compile_erase()  ← SIGSEGV
> ```
> 
> What's happening: PR #142 wraps each compiled while-loop body in a `compile_erase` callback held by a `shared_ptr` (the `CacheGuard`) so that when the primitive is destroyed, MLX gets told to drop the cached compiled function. At process exit, two static destructors race:
> 
> 1. MLX's global `CompilerCache` destructor walks its hash table and tears down each cached `Compiled` node, which transitively destroys the `mlx::core::array`s holding `WhileLoopPrimitive`s.
> 2. Each of those `~WhileLoopPrimitive` calls tries to invoke `compile_erase` on the *same* `CompilerCache` that's mid-destruction.
> 
> PR #142 has an `IsProcessShuttingDown()` guard meant to short-circuit this exact case, but it isn't catching the path triggered from inside the cache's own destructor — the guard flag is set by `atexit`, but `__cxa_finalize_ranges` runs static destructors *before* atexit handlers fire, so the guard is still false when we reach it. Hence the segfault.
> 
> It's harmless for badge math (pytest had already exited 0 and written results), but it's a real bug — and it might also be what's causing the intermittent hangs we just saw, if the same race is occurring during a test's subprocess teardown rather than process exit. Worth flagging on #142.

**Comment by tillahoffmann** [comment]
@tsumme1, thanks for bearing with me! The hang was indeed unrelated to your changes. It's a weird ordering bug where scatter updating an element of an array that is NaN leads to a deadlock in MLX. Upstream fix at ml-explore/mlx#3492. The completion with non-zero exit code was a cleanup race (cf. #146 and #148).

Just running the last tests and then we're ready to merge.

**Comment by tillahoffmann** [comment]
Would be great to get your eyes on this @tsumme1 because you have a deeper understanding of how the primitives work.

**Comment by tsumme1** [comment]
@tillahoffmann Everything looks good to me. After your latest changes to this PR, I am able to run the tests with 92.5% passing, and my performance benchmark is still showing an improvement.

**Comment by tillahoffmann** [comment]
Excellent, thank you for confirming!
